docs/RFCS/data_structures_for_logical_planning.md

```diff
@@ -0,0 +1,1001 @@
+- Feature Name: Data structures for SQL logical plans
+- Status: draft
+- Start Date: 2010-10-12
+- Authors: knz, Peter Mattis
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: (one or more # from the issue tracker)
+
+# Summary
+
+This RFC *outlines* new data structures and principles around the
+handling of SQL logical plans, that both 1) lower the cost of managing
+logical plans in the current architecture 2) facilitate further work
+on SQL optimizations. It also proposes a *path to the introduction* of
+these data structures in the current CockroachDB source base.
+
+The new data structures are:
+
+- *classes* to represent nodes in the SQL expression tree, where a
+  single class can contain multiple alternative equivalent expressions
+  that share a common set of properties, and where individual
+  expressions refer to entire classes as children/siblings rather than
+  other expressions.
+- a *memo* data structure that stores all the classes. The name
+  "memo" is used because it memoizes the results of analyzing/generating
+  alternative plans for a single query.
+
+The proposed path is to make this new code co-exist with the current code.
+In comparison with the current code in CockroachDB, the main change is
+the introduction of expression classes, so that expressions do not
+form a "textbook" tree in memory any more.
+
```

Comment:
Thanks for writing this. Is it possible for you to describe some problems that this new proposal is fixing? The summary discussing the introduction of two new data structures but in reading it it's really hard to understand at that point in reading that the particular choice is valid/optimal. It also doesn't provide any historical context. Thanks a lot!